### PR TITLE
Add SaveChunkEvent.

### PR DESCRIPTION
**Pull request by Zidane**
For discussion, @dualspiral .

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

### CODE REVIEW

File: src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import org.spongepowered.api.world.Chunk;
+
+/**
+ * Called when a {@link Chunk} is performing a save.
+ */
+public interface SaveChunkEvent extends Event {
+
+    Vector3i getPosition();
+
+    interface Pre extends SaveChunkEvent, TargetChunkEvent, Cancellable {}
```
Comment (dualspiral) [inline src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java:48]: Why might you want to cancel a chunk save? Isn't that what the serialisation behaviours for?

File: src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import org.spongepowered.api.event.Cancellable;
+import org.spongepowered.api.event.Event;
+import org.spongepowered.api.world.Chunk;
+
+/**
+ * Called when a {@link Chunk} is performing a save.
+ */
+public interface SaveChunkEvent extends Event {
+
+    Vector3i getPosition();
```
Comment (dualspiral) [inline src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java]: I'm assuming that this is the chunk position - in which case I think we should name this `getChunkPosition()` to make it crystal clear that's what this is. I know we'll add javadocs to this BUT many a developer may not read the javadocs and assume.

File: src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    interface Pre extends SaveChunkEvent, TargetChunkEvent, Cancellable {}
+
+    interface Post extends SaveChunkEvent {}
```
Comment (dualspiral) [inline src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java:56]: Just read Discord, this basically means that the save is going ahead, but may not yet be on the disc. We will want to mention that in the JDs.

**Review by dualspiral — COMMENTED** [review]
Discussion, as requested. Just trying to get a feel for what everything is about.

File: src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java (hunk 2 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java]: Agreed

File: src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java (hunk 1 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java:48]: SerializationBehavior is all or nothing. You could choose to not save certain chunks. Debate could be made either way

File: src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java (hunk 3 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java:56]: It was but ultimately that isnt enough for Joni's usecase. He needs to be sure that file wrote to disk. Therefore this is as post as you get, the file is on the disk.

File: src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java (hunk 1 above)
Comment (dualspiral) [reply src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java:48]: I don't have an issue with this personally - dunno if anyone else does.

Comment (Zidane) [reply src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java:48]: One thought I had in mind is an area of the world where any changes wouldn't be wrote. You enter this "zone" and when you leave and it unloads and loads again the zone resets.

Think "mining worlds" but no need to do more worlds, lol.

**Comment by dualspiral** [comment]
For note, associated PR: https://github.com/SpongePowered/SpongeCommon/pull/2551

Is this good to go, or is there more to consider? Anyone else have any comments on this?

File: src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java (hunk 1 above)
Comment (gabizou) [reply src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java:48]: I can see some issues with this though, if handled improperly, couldn't the chunk load having not been saved cause some entity weirdness?

File: src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+     */
+    interface Pre extends SaveChunkEvent, TargetChunkEvent, Cancellable {}
+
+    /**
+     * Called after the {@link Chunk} is saved. Guaranteed to exist in the chunk's
+     * storage container.
+     *
+     * Depending on the implementation, this event may be called off-thread.
+     */
+    interface Post extends SaveChunkEvent {}
```
Comment (ImMorpheus) [inline src/main/java/org/spongepowered/api/event/world/chunk/SaveChunkEvent.java:56]: Shouldn't this `extends TargetChunkEvent` ?

EDIT: no, it's too late

**Review by ImMorpheus — APPROVED** [review]

**Review by gabizou — APPROVED** [review]
LGTM, Can always refigure this in API 8 due to the nature of the chunk management being redone there.

**Pull request merged**